### PR TITLE
Add CI Testing Section

### PR DESCRIPTION
**Pull request by gramss**
Hey,

I just explored CI testing for myself and had the feeling of writing about what the spirit of CI testing is and what entry points for CI beginners might be interesting.

Also included is my example of a working partial script of RewrittenYaml.

Hope this fits into the documentation?

Cheers

### CODE REVIEW

File: tutorials/docs/ci_testing.rst [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+**Overview:**
+
+- `Benefits of Automated Testing`_
+- `Hands on Testing`_
+- `Rewriting Parameter Values from YAML Files in Launch scripts`_
+
+Benefits of Automated Testing
+=============================
+
+.. image:: images/CI_Testing/code_coverage_bump_40_80.png
```
Comment (SteveMacenski) [inline tutorials/docs/ci_testing.rst:14]: What's the point this image is trying to convey?

File: tutorials/docs/ci_testing.rst [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    :align: center
+
+Whenever a feature gets added or altered, it is best practice to confirm the intended outcome with automated test scripts.
+This document features an easy step by step guide how CI should be integrated alongside new PRs.
+
+Also, not only new code should have CI testing. To find nasty bugs and corner cases, it is Nav2's ultimate goal to have a test coverage of nearly 100% !
```
Comment (SteveMacenski) [inline tutorials/docs/ci_testing.rst:20]: .... don't make promises for me ;-) 

Lets change that to 90%. That's my goal for every individual package.

File: tutorials/docs/ci_testing.rst [hunk 3]
```diff
@@ -0,0 +1,224 @@
+.. _ci_testing:
```
Comment (SteveMacenski) [inline tutorials/docs/ci_testing.rst:1]: So the big things here are:
- Audience: who is this document for? 
- Goals: What is this document trying to accomplish?

Its not clear to me what this is trying to do. I don't disagree having some content about testing in documentation, but I'm not sure what the aim or audience this document has.

"CI" seems odd here too. You mostly talk about testing, almost nothing about CI beyond the code coverage report.

File: tutorials/docs/ci_testing.rst [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+Before you can test anything, make sure you have the basic examples of the :ref:`getting_started` guides up and running.
+
+For integrating and running your own tests, you have to build nav2 from source. Using nav2 main with ros2 installed as binaries technically works, 
+but automated CI testing happening on github when you merge you create a PR for the nav2 stack also uses ros2 rolling (= latest/main).
+Therefor, it is recommend to test your new code with the nav2 stack based on the main branch and also let the script install ros2 from source.
+
+1. Run Existing Tests
+---------------------
+Nav2 specific tests can be found in each individual component under a dedicated test folder. 
+One additional package, ``nav2_system_tests``, `exists <https://github.com/ros-planning/navigation2/tree/main/nav2_system_tests>`_ for testing components, sub-systems and full system tests.
```
Comment (SteveMacenski) [inline tutorials/docs/ci_testing.rst:39]: ", in addition to unit and integration tests in individual packages."

File: tutorials/docs/ci_testing.rst [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+This package also comes with a ``README.md`` pointing out, that unit tests for sub-functions of components should be provided within each individual component.
+
+To run existing tests of components or the whole system from within the ``nav2_system_tests`` package, you can use this procedure:
+
+  - build nav2 with ``$ colcon build --symlink-install``
+  - navigate to the component you would like to test ``$ cd build/<package_name>``
```
Comment (SteveMacenski) [inline tutorials/docs/ci_testing.rst:45]: This isn't good. You can just do `colcon test --packages-select <pkg-name>` to run the tests of an individual package. You should never need to go into `build`.

File: tutorials/docs/ci_testing.rst [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+  - run all the tests of this package ``$ ctest -V``
+
+To run all tests of all nav2 components, you can also use ``colcon`` for convenience:
```
Comment (SteveMacenski) [inline tutorials/docs/ci_testing.rst:48]: Doesn't mention linters, which also run if its enabled in the `if testing` `CMakeLists`.

Comment (SteveMacenski) [reply tutorials/docs/ci_testing.rst:48]: Linters in the blob below wont just run by themselves, it has to be in the cmakelists or invoked manually.

E.g. you might want to have a subsection about running individual liners on CLI `ament_{cpplint, cppcheck, uncrustify, flake8, etc}`

File: tutorials/docs/ci_testing.rst [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+2. Run Individual Tests
+-----------------------
+Tests are declared in individual ``CMakeList.txt`` files in the test sub-folders of components or in ``nav2_system_tests``.
+While developing individual tests, it might make sense to only run one out of multiple tests.
+This can be done with either:
+
+- ``$ ctest -V -R <regex>`` in the respective folder **or**
+- ``$ colcon test --packages-select <name-of-pkg> --ctest-args -V -R <regex>``
+
+Where ``regex`` represents the name or search-expression for the test(s) you want to run manually.
```
Comment (SteveMacenski) [inline tutorials/docs/ci_testing.rst:86]: Some example regex for a common thing would be nice here.

File: tutorials/docs/ci_testing.rst [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+- Expect things to break: this is good! -> make sure to catch all errors and handle them accordingly
+- Combine your components with other test: If feasible create corner scenarios where your code really should improve things
+- Quantity over complexity: Better write multiple tests than make them to complicate so others cannot understand why it fails in the future
+
+**What tools do I have?**
+
+- look and learn from existing code in the nav2 stack, we have plenty of tests!
+- play with different launch parameters: Have a look at overwriting them in the next section
+
+4. Register Your Own Test
```
Comment (SteveMacenski) [inline tutorials/docs/ci_testing.rst:112]: I don't think "register" is the word you want. That's a plugin-y word that doesn't have use in Cmakel

File: tutorials/docs/ci_testing.rst [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+
+Make sure to include your new Tests in the specific ``CMakeList.txt`` file and recompile your working-space with ``colcon build --symlink-install``.
+Depending on writing tests in C++ or python there are different ways to register tests.
+
+This first example is for registering tests surrounding behavior tree actions.
+`Source c++ test <https://github.com/ros-planning/navigation2/blob/main/nav2_behavior_tree/test/plugins/action/CMakeLists.txt>`_ 
+
+.. code-block:: text
+
+  ament_add_gtest(test_action_spin_action test_spin_action.cpp)
```
Comment (SteveMacenski) [inline tutorials/docs/ci_testing.rst:123]: Please explain these cmake macros and what each do

File: tutorials/docs/ci_testing.rst [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+  ament_target_dependencies(test_action_back_up_action ${dependencies})
+
+
+Here is an example for testing with python and pytest, especially useful for testing launch sequences.
+Interesting to note are the free set-able environment variables that can later be used to rewrite parameter values for launch scripts.
+`Source python test <https://github.com/ros-planning/navigation2/blob/main/nav2_system_tests/src/system/CMakeLists.txt>`_
+
+.. code-block:: text
+
+  ament_add_test(test_bt_navigator_with_groot_monitoring
```
Comment (SteveMacenski) [inline tutorials/docs/ci_testing.rst:138]: This isn't just for python, this is for running any launch-based test vs a raw `gtest` test. So all these examples are used to run launch files which have test nodes that are launched as well as python. But they dont need to be python.

File: tutorials/docs/ci_testing.rst [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+5. Check Your Test with CI and Check Code Coverage
+--------------------------------------------------
+
+.. image:: images/CI_Testing/github_coverage_diff.png
+    
+
+Is there room to improve the test coverage near your code? Ain't you just the right expert about this code section?
+Think about adding tests that exceed your own focus and help improve nav2/ros2 reach a higher overall code coverage and ultimately also quality.
+
+The report above is an automated post by codecov.io-bot on github that posts results of CI automatically for every new PR.
```
Comment (SteveMacenski) [inline tutorials/docs/ci_testing.rst:164]: I'd also recommend seeing the report and look at the files you modified to see the specific lines that were / weren't covered (or see the new package total %). That can help you see where you need to add more tests for or if you hit your test coverage goal (sometimes there's noise in just that high level % from other packages moving around a little).

File: tutorials/docs/ci_testing.rst [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+Please consider helping increase the code coverage and use the opportunity to learn more about the internals of the navigation2 stack! 
+
+Rewriting Parameter Values from YAML Files in Launch scripts
+============================================================
+In most occasions some small new features are added and made available through a few new parameters. As the standard nav2 user should not be overloaded with features it makes sense to disable most of the additional or drop-in features in the default ``params.yaml`` file.
```
Comment (SteveMacenski) [inline tutorials/docs/ci_testing.rst:169]: The explanation here doesn't 100% explain why this is in this document.

File: tutorials/docs/ci_testing.rst [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+- ``$ ctest -V -R <regex>`` in the respective folder **or**
+- ``$ colcon test --packages-select <name-of-pkg> --ctest-args -V -R <regex>``
+
+Where ``regex`` represents the name or search-expression for the test(s) you want to run manually. 
+You can also find the name of a certain test by running all tests of the desired package. 
+
+.. note::
+  When testing with ``pytest`` - typically for ros2 launch files -  and building your package with ``$ colcon build --symlink-install``, you can even change your test scripts without rebuilding the whole package! 
+
+3. Writing Your Own Test
```
Comment (SteveMacenski) [inline tutorials/docs/ci_testing.rst:92]: Explaining gtest would be good. Or links to launch test uses

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
See second comment: I think this needs a better understanding of audience and goal. I'm not sure what the reader walks away from this with or who its targeted at.

File: tutorials/docs/ci_testing.rst (hunk 1 above)
Comment (gramss) [reply tutorials/docs/ci_testing.rst:14]: Basically "stonks" (the meme).

A growing coverage should motivate others to help raise the bar to 100 (or as you more realistically said: 90%).
And they should question themselves what it takes to create such a huge bump from 40 to ~75.
Also, they should see that it can be big jumps and also small jumps (new tests added) that also help improve.

So basically multiple things, that also motivated me to write something about it..

Disclaimer: I'm absolutely new to CI / testing. I just wrote down my understanding of things while I discovered them. So thanks for the input! I'll try to improve the points you mentioned and learn a few things more on the go.

File: tutorials/docs/ci_testing.rst (hunk 2 above)
Comment (gramss) [reply tutorials/docs/ci_testing.rst:20]: sure thing.. :-) done.

File: tutorials/docs/ci_testing.rst (hunk 3 above)
Comment (gramss) [reply tutorials/docs/ci_testing.rst:1]: Audience:
Get new folks involved with testing. So programmers that already added some code and want to nicely finish their PR with tests but have no clue where to even start looking in a massive project like nav2.
Testing does not take ages to figure it out by themself, but it should be just something that gives you some good starting points and get to the point and continue programming

Goal:
Make clear to them that testing is not hard and can help you to get to know your code better

CI and testing where synonyms for me at the beginning of this document. Testing in CI is also just a little chapter of CI.
Probably should remove CI references in the file names..
But changed the setting now a bit. Let me hear what you think.

File: tutorials/docs/ci_testing.rst (hunk 4 above)
Comment (gramss) [reply tutorials/docs/ci_testing.rst:39]: done

File: tutorials/docs/ci_testing.rst (hunk 5 above)
Comment (gramss) [reply tutorials/docs/ci_testing.rst:45]: yes you can do that.
But it is a really really long command this way..:
```
colcon test --event-handlers console_direct+ --packages-select <pkg-name> --ctest-args -R <regex>
```
I especially want to see the output via the event-handlers as I need to see what is happening with my new test. The new test mostly is not right the first time. (during normal test times, this should show nothing as it is currently)

Probably still better to use colcon than go to the build folder.

Done.

File: tutorials/docs/ci_testing.rst (hunk 7 above)
Comment (gramss) [reply tutorials/docs/ci_testing.rst:86]: done

File: tutorials/docs/ci_testing.rst (hunk 8 above)
Comment (gramss) [reply tutorials/docs/ci_testing.rst:112]: changed to "add"

File: tutorials/docs/ci_testing.rst (hunk 10 above)
Comment (gramss) [reply tutorials/docs/ci_testing.rst:138]: Changed it.
I thought that launch scripts in ros2 must always be python

**Comment by gramss** [comment]
there are 4 uncommented requests from you that I will come back shortly.

File: tutorials/docs/ci_testing.rst (hunk 5 above)
Comment (SteveMacenski) [reply tutorials/docs/ci_testing.rst]: You're talking about in this section about running "existing tests of components", so you should target that audience. Not the "run a single new test" audience, these are different things.

File: tutorials/docs/ci_testing.rst (hunk 1 above)
Comment (SteveMacenski) [reply tutorials/docs/ci_testing.rst:14]: Lets not use this image, its deceptive, its a little more complex to describe why that jump happened, but its not as obvious as you think

File: tutorials/docs/ci_testing.rst (hunk 3 above)
Comment (SteveMacenski) [reply tutorials/docs/ci_testing.rst:1]: So why is it called "CI testing" if its meant for new people learning about testing?

I'm not sure this really hits the target on helping people get into testing. There's not really a linear narrative explaining concepts, showing how to use tools, and helping write tests. There's no real mention or details on gtest and test-launch and the ament testing tools, which I'd think would be necessary for a "how do I setup tests" tutorial. There's not any code to be a "writing my first test" tutorial. There's not really a clear direction this content is going in

File: configuration/packages/configuring-bt-navigator.rst [hunk 14]
```diff
@@ -53,6 +53,43 @@ Parameters
...
+    Consider checking out the :ref:`groot_introduction` tutorial for using this feature. 
+
+    **Attention:** Groot will only work after the behavior tree is running, which means that the nav2 stack has to receive a goal first.
+
+:groot_zmq_publisher_port:
+
+  ====== ======= ======= ========
+  Type   Default Unit    Optional
+  ------ ------- ------- --------
+  int    1666    Port    yes
```
Comment (SteveMacenski) [inline configuration/packages/configuring-bt-navigator.rst:76]: Remove these optional fields, any parameter with a default is technically optional. As well with unit, please stick with the format unless you'd like to update all of the parameters in all of the guides to this new format :wink:

File: tutorials/docs/groot.rst [hunk 15]
```diff
@@ -0,0 +1,88 @@
+.. _groot_introduction:
```
Comment (SteveMacenski) [inline tutorials/docs/groot.rst:1]: Groot tutorial and deleted the CI testing tutorial? I'm confused, this is the CI testing section PR.

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
I think you pushed your groot stuff to your CI testing PR branch, then deleted your CI testing work.

**Comment by SteveMacenski** [comment]
Please revert all groot stuff from this PR and open a new PR with groot docs
